### PR TITLE
separate e2e envs & enable database assert on test env

### PR DESCRIPTION
**Pull request by axle-h**

### CODE REVIEW

File: src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.justice.digital.hmpps.deliusapi.config
+
+import org.springframework.test.context.ActiveProfilesResolver
+
+class EndToEndTestActiveProfilesResolver : ActiveProfilesResolver {
+  override fun resolve(testClass: Class<*>): Array<String> {
+    val env = System.getenv("SPRING_PROFILES_ACTIVE")?.split(',', ' ')?.toTypedArray()
```
Comment (marcus-bcl) [inline src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt:7]: Could this be achieved by adding `spring.profiles.include: e2e` to the config files?

**Review by marcus-bcl — APPROVED** [review]

File: src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt (hunk 1 above)
Comment (electrified) [reply src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt]: I wondered about this too - is it so you can optionally run against different DBs by setting additional profiles using the env var, without modifying any of the committed files?

File: src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/v1/contact/CreateContactV1Test.kt [hunk 2]
```diff
@@ -23,13 +25,15 @@ class CreateContactV1Test @Autowired constructor(
   private lateinit var created: Contact
 
   @Test
+  @Transactional
```
Comment (electrified) [inline src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/v1/contact/CreateContactV1Test.kt:28]: Did the test not work without the transaction annotation?

For mine I think I will have to leave this off, as test transactions roll back, which will revert the deletion.

**Review by electrified — APPROVED** [review]

File: src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt (hunk 1 above)
Comment (axle-h) [reply src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt]: afraid not, here's the default: https://github.com/spring-projects/spring-framework/blob/master/spring-test/src/main/java/org/springframework/test/context/support/DefaultActiveProfilesResolver.java
Makes sense to me for unit & integration tests TBF as you would have all the environment configu hard coded.

File: src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/v1/contact/CreateContactV1Test.kt (hunk 2 above)
Comment (axle-h) [reply src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/v1/contact/CreateContactV1Test.kt:28]: :-(
This is required to have a hibernate session present for lazy properties to work. There may be an alternative to this, like wrapping the database assert bit only in an explicit transaction.

File: src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt (hunk 1 above)
Comment (axle-h) [reply src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt]: @electrified yeah, this is meant for selecting which environment to use, which all have different reference data. This reminds me that the buildspec.yml should be updated to activate the `test` profile.

Comment (axle-h) [reply src/e2e/kotlin/uk/gov/justice/digital/hmpps/deliusapi/config/EndToEndTestActiveProfilesResolver.kt]: Just pushed a fix. btw there is documentation for this in: `doc/testing.md`

**Pull request merged**